File: frontend/frontend/src/app/services/clan-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { ClanDto } from '../common/clan/clan-dto';
import { Observable, of } from 'rxjs';
import { CreateClanPayload } from '../common/clan/create-clan-payload';
import { ClanLeaderboardDto } from '../common/clan/clan-leaderboard-dto';
import { catchError } from 'rxjs';
import { ClanMemberContributionDto } from '../common/clan/clan-member-contribution-dto';
import { UpdateClanPayLoad } from '../common/clan/update-clan-pay-load';
import { ClanMemberDto } from '../common/clan/clan-member-dto';

@Injectable({
  providedIn: 'root',
})
export class ClanService {
   private readonly baseUrl = '/api/clans';

  constructor(private httpClient: HttpClient) {}

  getMyClan(): Observable<ClanDto | null> {
    return this.httpClient.get<ClanDto>(`${this.baseUrl}/me`).pipe(
      catchError((error) => {
        if (error.status === 404 || error.status === 204) {
          return of(null);
        }

        console.error('Failed to load my clan:', error);
        return of(null);
      })
    );
  }

  updateClan(
  clanId: number,
  payload: UpdateClanPayLoad
): Observable<ClanDto> {
  return this.httpClient.put<ClanDto>(
    `${this.baseUrl}/${clanId}`,
    payload
  );
}

  getClanById(clanId: number): Observable<ClanDto> {
    return this.httpClient.get<ClanDto>(`${this.baseUrl}/${clanId}`);
  }

  getTopClans(): Observable<ClanLeaderboardDto[]> {
    return this.httpClient.get<ClanLeaderboardDto[]>(`${this.baseUrl}/top`);
  }

  createClan(payload: CreateClanPayload): Observable<ClanDto> {
    return this.httpClient.post<ClanDto>(this.baseUrl, {
      name: payload.name,
      description: payload.description ?? '',
      invite: payload.invite ?? false,
    });
  }

  joinClan(clanId: number): Observable<void> {
    return this.httpClient.post<void>(`${this.baseUrl}/${clanId}/join`, {});
  }

  leaveClan(clanId: number): Observable<void> {
    return this.httpClient.delete<void>(`${this.baseUrl}/${clanId}/leave`);
  }

  getClanContributions(clanId: number): Observable<ClanMemberContributionDto[]> {
  return this.httpClient.get<ClanMemberContributionDto[]>(
    `${this.baseUrl}/${clanId}/contributions`
  );
}

getPendingRequests(clanId: number): Observable<ClanMemberDto[]> {
  return this.httpClient.get<ClanMemberDto[]>(
    `${this.baseUrl}/${clanId}/pending`
  );
}

acceptPendingMember(clanId: number, targetUserUuid: string): Observable<void> {
  return this.httpClient.post<void>(
    `${this.baseUrl}/${clanId}/pending/${targetUserUuid}/accept`,
    {}
  );
}

declinePendingMember(clanId: number, targetUserUuid: string): Observable<void> {
  return this.httpClient.post<void>(
    `${this.baseUrl}/${clanId}/pending/${targetUserUuid}/decline`,
    {}
  );
}

kickMember(clanId: number, targetUserUuid: string): Observable<void> {
  return this.httpClient.delete<void>(
    `${this.baseUrl}/${clanId}/kick?targetUserId=${targetUserUuid}`
  );
}

transferLeadership(clanId: number, newLeaderUserUuid: string): Observable<void> {
  return this.httpClient.post<void>(
    `${this.baseUrl}/${clanId}/leadership/transfer?newLeaderUserId=${newLeaderUserUuid}`,
    {}
  );
}
}

File: backend/src/main/java/com/strongBeton/strongBeton/service/clan/ClanService.java
package com.strongBeton.strongBeton.service.clan;

import com.strongBeton.strongBeton.dto.clan.ClanDTO;
import com.strongBeton.strongBeton.dto.clan.ClanMemberContributionDTO;
import com.strongBeton.strongBeton.dto.clan.ClanMemberDTO;
import com.strongBeton.strongBeton.entity.user.User;

import java.util.List;
import java.util.UUID;

public interface ClanService {

    ClanDTO createClan(ClanDTO clanDTO, User user);
    ClanDTO getMyClan(User user);
    ClanDTO getClanById(int clanId);
    ClanDTO updateClan(int clanId, User user, ClanDTO clanDTO);
    void deleteClan(int clanId, UUID userId);

    void joinClan(int clanId, User user);
    void leaveClan(int clanId, User user);
    void kickMember(int clanId, UUID targetUserId, User requester);
    void inviteMember(int clanId, UUID targetUserId, UUID requesterId);
    void acceptInvite(int clanId, UUID requesterId, int targetId);
    void declineInvite(int clanId, UUID requesterId, int targetId);

    void promoteMember(int clanId, int targetUserId, UUID requesterId);
    void demoteMember(int clanId, int targetUserId, UUID requesterId);
    public void transferLeadership(int clanId, UUID newLeaderUserId, User requester);

    List<ClanMemberDTO> getPendingRequests(int clanId, User requester);

    void acceptPendingMember(int clanId, UUID targetUserUuid, User requester);

    void declinePendingMember(int clanId, UUID targetUserUuid, User requester);




}
